crates/typos-cli/src/bin/typos-cli/report.rs

```diff
@@ -280,6 +285,172 @@ impl Report for PrintJson {
     }
 }
 
+#[derive(Debug)]
+pub(crate) struct PrintSarif {
+    results: Mutex<Vec<sarif::Result>>,
+    error: Mutex<Vec<String>>,
+}
+
+impl Default for PrintSarif {
+    fn default() -> Self {
+        Self {
+            results: Mutex::new(Vec::new()),
+            error: Mutex::new(Vec::new()),
+        }
+    }
+}
+
+impl Report for PrintSarif {
+    fn report(&self, msg: Message<'_>) -> Result<(), std::io::Error> {
+        self.report_sarif(msg).map_err(sarif_error_mapper)
+    }
+
+    fn generate_final_result(&self) -> Result<(), std::io::Error> {
+        self.generate_final_result().map_err(sarif_error_mapper)
+    }
+}
+
+impl PrintSarif {
+    fn report_sarif(&self, msg: Message<'_>) -> Result<(), Box<dyn std::error::Error>> {
+        match &msg {
+            Message::Typo(msg) => {
+                let message = match &msg.corrections {
+                    typos::Status::Valid => {
+                        return Ok(());
+                    }
+                    typos::Status::Invalid => {
+                        format!("`{}` is disallowed", msg.typo)
+                    }
+                    typos::Status::Corrections(corrections) => {
+                        format!(
+                            "`{}` should be {}",
+                            msg.typo,
+                            itertools::join(corrections.iter().map(|s| format!("`{s}`")), ", ",)
+                        )
+                    }
+                };
+
+                let start = String::from_utf8_lossy(&msg.buffer[0..msg.byte_offset]);
+                let column_start = start.encode_utf16().count() + 1;
+                let column_end = msg.typo.encode_utf16().count() + column_start;
+                let line_num = msg
+                    .context
+                    .as_ref()
+                    .map(|context| match context {
+                        Context::File(context) => context.line_num,
+                        _ => 1,
+                    })
+                    .unwrap_or(1);
```

Comment:
I'm a little concerned about how we create locations.

Could we mirror what `clippy-sarif` does?
See https://github.com/psastras/sarif-rs/blob/main/serde-sarif/src/converters/clippy.rs

(note that also has suggestion support)

Reply:
suggestion support has been implement